add action node

KeeyanGhoreshi opened the pull request:
i think this gets the gist.  Add an action node, get an action node.

blangley28 commented on src/components/Sidebar/ActionNode.tsx:
```diff
@@ -1,49 +1,212 @@
+            title: ''
+          }
+        };
+      } else if (
+        currentState.action.length > 0 &&
+        currentState.action[0].resource.resourceType !== 'Careplan'
+      ) {
+        action = currentState.action[0];
+        action.resource.resourceType = actionType?.label;
```
Did the CarePlan replace the need for multiple actions? If so then we should probably update the data model to make Action a single object instead of a list.
The reason I bring this up is because it looks like this block means we cannot add multiple actions - it just replaces the action type of the first one

dehall replied:
I told Keeyan separately that since we only ever show one action at a time, for starters this PR should use the existing data model and only put one item in the list. If we want to change the data model, which I would be good with, let's do that separately.

KeeyanGhoreshi replied:
I'm not sure what to do here since the designs only show one action being added but the data model allows for multiple.  I've talked to dylan and we decided to keep the data model the same even though it only allows 1 action for now.

blangley28 commented on src/components/Sidebar/ActionNode.tsx:
```diff
@@ -1,49 +1,212 @@
+      if (!currentNodeKey) return;
+
+      const code = event?.target.value || '';
+      const currentState: GuidanceState = pathway.states[currentNodeKey] as GuidanceState;
+      const action: Action = currentState.action[0];
```
We should probably put something in the code style guide to keep everything consistent but you don't need the types here
```suggestion
      const currentState = pathway.states[currentNodeKey] as GuidanceState;
      const action = currentState.action[0];
```

blangley28 commented on src/components/Sidebar/ActionNode.tsx:
```diff
@@ -1,49 +1,212 @@
   const changeNodeTypeEnabled = currentNode.key !== undefined && currentNode.key !== 'Start';
 
+  const action = currentNode.action;
+  const resource = action && action.length > 0 && action[0].resource;
```
The null check for action can be added to the second statement
```suggestion
  const resource = action?.length > 0 && action[0].resource;
```

blangley28 commented on src/components/Sidebar/ActionNode.tsx:
```diff
@@ -1,49 +1,212 @@
+                  value={resource.code.coding[0].code}
+                />
+              )}
+              {resource.code.coding[0].system && (
```
If this was intended to be hidden until a system is selected then is there a reason it can't be included with the `Input` above it? Perhaps it should be hidden until a code is set though

blangley28 commented on src/components/Sidebar/ActionNode.tsx:
```diff
@@ -1,49 +1,212 @@
     },
     [changeNodeType]
   );
+  const currentNodeKey = currentNode?.key;
```
We should be consistent with using `currentNode` and its key:
1. For starters I don't think `currentNode` can be null so we shouldn't need to check for that
2. This also means all of the `if (!currentNodeKey) return;` statements can be removed
3. The `currentNode` is already passed as a prop but in multiple places you use the key to find the node from the pathway instead of just using the node passed in.
4. Some places use `currentNodeKey` but others use `currentNode.key`. I don't really mind which one we use but it should be consistent.

KeeyanGhoreshi replied:
currentNodeKey can be undefined, so we still have to check it.  Other stuff I agree.

blangley28 commented on src/types/pathways-model.d.ts:
```diff
@@ -37,19 +37,20 @@ declare module 'pathways-model' {
     cql: string;
     elm?: ElmLibrary;
     action: Action[];
+    title?: string;
+    description?: string;
```
What are these used for?

blangley28 commented on src/utils/builder.ts:
```diff
@@ -268,6 +268,32 @@ export function setStateCriteriaSource(
+  };
+}
+
+export function setStateCodeSystem(pathway: Pathway, key: string, action: Action[]): Pathway {
+  return {
+    ...pathway,
+    states: {
+      ...pathway.states,
+      [key]: {
+        ...pathway.states[key],
+        action
+      }
+    }
+  };
+}
+
```
This method is identical to the one above

blangley28 commented on src/utils/builder.ts:
```diff
@@ -268,6 +268,32 @@ export function setStateCriteriaSource(
   };
 }
 
+export function setStateAction(pathway: Pathway, key: string, action: Action[]): Pathway {
```
When a guidance state is created it initialized the action to be an empty list. I suggest making this method `addStateAction` to create a new action. Then use the other set action methods to set the specific attributes of the action

KeeyanGhoreshi replied:
isn't this already accomplished inside the action node?  What benefit do we get from swapping stuff around like this?

blangley28 commented:
When I select an action node is doesn't pre-populate the resource (action) type
![Screen Shot 2020-06-15 at 12 15 08 PM](https://user-images.githubusercontent.com/53485517/84681263-dd3fde80-af01-11ea-91ea-8ce7647bfd78.png)
But once I manually select the correct type it populates the coding
![Screen Shot 2020-06-15 at 12 16 17 PM](https://user-images.githubusercontent.com/53485517/84681370-05c7d880-af02-11ea-8e3b-d851fa7296cb.png)

blangley28 commented on src/components/Sidebar/ActionNode.tsx:
```diff
@@ -1,49 +1,212 @@
 ];
 
+const actionTypeOptions = [
+  { label: 'Medication Request', value: 'Medication Request' },
+  { label: 'Service Request', value: 'Service Request' },
+  { label: 'Careplan', value: 'Careplan' }
```
The values here are being used in the `resource` but are not the correct fhirnames
```suggestion
  { label: 'Medication Request', value: 'MedicationRequest' },
  { label: 'Service Request', value: 'ServiceRequest' },
  { label: 'Careplan', value: 'CarePlan' }
```

blangley28 commented on src/components/Sidebar/ActionNode.tsx:
```diff
@@ -1,49 +1,212 @@
+      const code = event?.target.value || '';
+      const currentState: GuidanceState = pathway.states[currentNodeKey] as GuidanceState;
+      const action: Action = currentState.action[0];
+      action.resource.code.coding[0].code = code;
```
The code isn't always `resource.code.coding[0]`. For `MedicationRequest` its `resource.medicationCodeableConcept`

noranda commented:
Designs for description field:
https://invis.io/6TX9T30J3M4#/418719440_Camino-Builder-Action-05

blangley28 commented:
I don't think this is an issue with just PR but I wanted to bring it up anyway - selecting from autofill in the text box looks really ugly. 
![Screen Shot 2020-06-17 at 1 42 30 PM](https://user-images.githubusercontent.com/53485517/84931155-6725ae00-b0a0-11ea-9ee2-1f67cd90da06.png)

Also from me filling out this form a lot on my branch it would be nice if after filling in a code you could click the "Validate" button or enter. In my branch I updated so the description field is hidden until it is "validated"

dehall commented:
What about the Add Action/Branch buttons at the bottom of the sidebar? https://projects.invisionapp.com/share/6TX9T30J3M4#/screens/421770938
I'm willing to create a new item for that if it's a big addition, but if it's easy and we can work that in here that would be great.

blangley28 commented:
> What about the Add Action/Branch buttons at the bottom of the sidebar? 
I don't follow - what about them? Have them hidden until all of the information is added?

dehall commented:
Sorry I should have a) tested more before commenting and b) been more clear. On a node that has an existing outgoing transition, nothing shows up in the sidebar to add a new option or edit the current transition: 
![image](https://user-images.githubusercontent.com/13512036/84934908-faadad80-b0a5-11ea-95d6-ef9ed61a4660.png)

On a node that doesn't have existing transitions the buttons do show up as expected: 
![image](https://user-images.githubusercontent.com/13512036/84934990-19ac3f80-b0a6-11ea-85cf-44d37edf7b13.png)

I think ultimately what we want in that first case is something similar to the branch node transitions editor, but without the CQL criteria parts. We still want to allow multiple transitions for when there are multiple options to choose. Let's make that a separate task so we can include more designs on how transitions should work.

noranda commented:
Issues remaining:

1. When an action node is first selected (click on "add action node"), there should be no node icon. Only when a user selects a node type should the corresponding icon appear in the node header.

2. Is there a reason we took out ICD-9 and ICD-10 code systems? For reference, CDS Connect uses the following code systems:
```js
const codeSystemOptions = [
  { value: 'SNOMED', label: 'SNOMED', id: 'http://snomed.info/sct' },
  { value: 'ICD-9-CM', label: 'ICD-9-CM', id: 'http://hl7.org/fhir/sid/icd-9-cm' },
  { value: 'ICD-10-CM', label: 'ICD-10-CM', id: 'http://hl7.org/fhir/sid/icd-10-cm' },
  { value: 'NCI', label: 'NCI', id: 'http://ncimeta.nci.nih.gov' },
  { value: 'LOINC', label: 'LOINC', id: 'http://loinc.org' },
  { value: 'RXNORM', label: 'RXNORM', id: 'http://www.nlm.nih.gov/research/umls/rxnorm' },
  { value: 'Other', label: 'Other', id: null }
];
```

3. As soon as an action type is selected (ie. the node now includes metadata), the current node should expand in the graph to show this.

4. Remove the Input component as it is not used anywhere.

5. The Action Type in the sidebar no longer matches the Type in the node. Recommend we change the Type in the node to include the new user-friendly types. Ex. "Procedure (ServiceRequest)" or just "Procedure".

![image](https://user-images.githubusercontent.com/5418735/85157487-8dc32080-b229-11ea-8366-0ab85d08b23b.png)

6. Regarding @blangley28's comment on the autofill styles, add this to the theme file on line 112 to fix:

```js
...
MuiInputBase: {
    ...,
    input: {
      ...,
      '&:-webkit-autofill': {
        transitionDelay: '9999s',
        transitionProperty: 'background-color, color'
      }
    }
  }
...
```

7. Remove the top margin, bold text, and fix font size on the display text to match the design.

<img width="465" alt="Screen Shot 2020-06-19 at 12 52 04 PM" src="https://user-images.githubusercontent.com/5418735/85159712-dc71ba00-b22b-11ea-8bc6-5a20d605b298.png">

Design:
<img width="475" alt="Screen Shot 2020-06-19 at 12 53 43 PM" src="https://user-images.githubusercontent.com/5418735/85159787-eeebf380-b22b-11ea-8e81-38a60eb35fec.png">

KeeyanGhoreshi commented:
I think I addressed all of the points in the latest commit(s).  Though for (3) I wasn't sure how exactly to get the Action Node to expand when it first got metadata.  Since assigning the type occurs in the `ActionNode`, and handling of expansion occurs in the `Node`, there's not really a good way to communicate to `Node` that a certain action has been fired in `ActionNode`.  

Since the action node can only really update the graph by changing the pathway's state, I just used the pathways state to check if the state changed from having no metadata to having some metadata.  But this solution seemed a bit hacky to me.  Ideally the node would expand as a result of the function in `ActionNode` populating the metadata, not as a side effect of checking the state every time to see if the metadata has been updated. 

I'm open to any ideas on how this could be improved, if there are any.

blangley28 commented:
Not sure if this is being addressed but I just rebased my branch off your latest changes I noticed the following
1. Clicking enter to validate doesn't work for regimens
2. Validating a regimen gives an error. 
>ActionNode.tsx:188 Uncaught TypeError: Cannot read property 'coding' of undefined
![Screen Shot 2020-06-22 at 6 49 15 PM](https://user-images.githubusercontent.com/53485517/85342621-14d9f800-b4b9-11ea-8790-b487f15ce529.png)

noranda commented on src/components/ExpandedNode/ExpandedNode.tsx:
```diff
@@ -28,10 +28,14 @@ type ExpandedNodeFieldProps = {
-      <td className={styles.desc}>{description}</td>
-    </tr>
+    <>
+      {description ? (
```
This can be better written as `{description && ( ... )}` if the else is null.

noranda commented:
Issues remaining:

1. An error is thrown when typing a Regimen title and then deleting it.

<img width="1032" alt="Screen Shot 2020-06-23 at 10 30 58 AM" src="https://user-images.githubusercontent.com/5418735/85416949-0f25f600-b53d-11ea-8437-ac4a138eaaf1.png">

2. I noticed for regimens/care plans, we are now inputing a title as opposed to a code and system, yet we are still displaying the option to validate. Are we able to validate this? If so, no changes needed. If not, we should remove the validate button.

<img width="464" alt="Screen Shot 2020-06-23 at 10 33 19 AM" src="https://user-images.githubusercontent.com/5418735/85417130-48f6fc80-b53d-11ea-8a9e-8ebdb6e70495.png">

noranda commented:
I found one last issue. To reproduce:

1. Create a new pathway
2. Add an action node
3. Expand the action node in the graph
4. Add a type to the action node

This will collapse the node which is not the behavior we want. The node should remain open, not toggle, when metadata exists. Everything else looks good to me.
